Gum-Joe/2Keys
# Development environment refactors

Some refactors for CU-5wt7m1 to tidy up the development environment:

- [x] Move all development dependencies to root
- [x] Use a root `nyc` which is extended, with configs moved to separate file
- [x] Use a single, extended `mocha` config
- [x] Clean up TSConfigs (to use projects references & incremental build -> one ts build command to compile everything)
- [x] Update CI to use more recent build methods (`wsrun`)
- [x] Use `test:all` script
- [x] Cache `.build-cache` for faster CI (REJECTED, causes build errors, and is not what should be done)

As much dev stuff should be controlled from the root of the repo.

Summary of changes (please copy into merge commit message):
- **MAJOR:** Move all development dependencies to the root of the repo, as a single source of truth
- **MAJOR:** `mocha` and `nyc` configs now at the root of the repo in `config/`, and extended by other modules
   - `nyc` config files: `nyc.config.js` and `.nycrc`
   - 'mocha` config files: `.mocharc.js`
   - There are config files in the root, these are for running tests and coverage on the entire rep at once (the `test:all` command)
- Add `all` option to NYC for coverage for all `src` files, not just those required (so we can see what files have np tests written for them)
   - Due to how the server is tested, the option is disabled ONLY for the `server` package
- **MAJOR NEW:** Moved to TS project references:
   - All modules' `tsconfig.json` extend `tsconfig.base.json` in the root
   - There's a single `tsconfig.json` in the root, used to build all the projects at once
   - **To build the repo now, used `yarn build` in the root of the repo.  This builds all the projects at once.**
      - Use `yarn build -v` to get logging output to see the order stuff is built.
   - & more.  **For all development-related instructions, please read the new `DEVELOPMENT.md` in the root of the repo (currently only found on this PR branch)**
- **NEW:** TS incremental builds now enabled, so only what is needed (new) to be built is built.
- Upgraded to ES2018 for TS compilation
- CI changes:
   - Use new build command
   - Use `coverage:all` command for coverage
